jni/src/faiss_index_service.cpp

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// The OpenSearch Contributors require contributions made to
+// this file be licensed under the Apache-2.0 license or a
+// compatible open source license.
+//
+// Modifications Copyright OpenSearch Contributors. See
+// GitHub history for details.
+
+#include "faiss_index_service.h"
+#include "faiss_methods.h"
+#include "faiss/index_factory.h"
+#include "faiss/Index.h"
+#include "faiss/IndexBinary.h"
+#include "faiss/IndexHNSW.h"
+#include "faiss/IndexBinaryHNSW.h"
+#include "faiss/IndexIVFFlat.h"
+#include "faiss/IndexBinaryIVF.h"
+#include "faiss/IndexIDMap.h"
+#include "faiss/index_io.h"
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <memory>
+#include <type_traits>
+
+namespace knn_jni {
+namespace faiss_wrapper {
+
+template<typename INDEX, typename IVF, typename HNSW>
+void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
+                        const std::unordered_map<std::string, jobject>& parametersCpp, INDEX * index) {
+    std::unordered_map<std::string,jobject>::const_iterator value;
+    if (auto * indexIvf = dynamic_cast<IVF*>(index)) {
+        if ((value = parametersCpp.find(knn_jni::NPROBES)) != parametersCpp.end()) {
+            indexIvf->nprobe = jniUtil->ConvertJavaObjectToCppInteger(env, value->second);
+        }
+
+        if ((value = parametersCpp.find(knn_jni::COARSE_QUANTIZER)) != parametersCpp.end()
+                && indexIvf->quantizer != nullptr) {
+            auto subParametersCpp = jniUtil->ConvertJavaMapToCppMap(env, value->second);
+            SetExtraParameters<INDEX, IVF, HNSW>(jniUtil, env, subParametersCpp, indexIvf->quantizer);
+        }
+    }
+
+    if (auto * indexHnsw = dynamic_cast<HNSW*>(index)) {
+
+        if ((value = parametersCpp.find(knn_jni::EF_CONSTRUCTION)) != parametersCpp.end()) {
+            indexHnsw->hnsw.efConstruction = jniUtil->ConvertJavaObjectToCppInteger(env, value->second);
+        }
+
+        if ((value = parametersCpp.find(knn_jni::EF_SEARCH)) != parametersCpp.end()) {
+            indexHnsw->hnsw.efSearch = jniUtil->ConvertJavaObjectToCppInteger(env, value->second);
+        }
+    }
+}
+
+IndexService::IndexService(std::unique_ptr<FaissMethods> faissMethods) : faissMethods(std::move(faissMethods)) {}
```

Comment:
The std::move is an interesting choice. But what my thinking is why we are using std::move here rather than just say a [unique pointer](https://en.cppreference.com/w/cpp/memory/unique_ptr) or [may do a call by reference](https://www.tutorialspoint.com/cplusplus/cpp_function_call_by_reference.htm) if we are concerned about copying of objects.

Reply:
Call by reference increase a chance of memory leak by bug.

Reply:
as this is an object, and you used the unique pointer, I doubt its the case. But if we 100% sure that this will not cause any performance issues and other memory leak I am okay to approve the code.

Reply:
This std::move is passing its ownership. If you don't use it, the ownership won't be transferred to IndexService and the memory can be released while IndexService sill needs it.